### ISSUE TITLE
Downgrade unsupported metadata record codes from errors to warnings

### ISSUE THREAD

**Issue by RyanGlScott** [GaloisInc/llvm-pretty-bc-parser#355]
We have made some strides towards making such that `llvm-pretty-bc-parser` doesn't emit a fatal error if you encounter a form of metadata that isn't currently supported (see https://github.com/GaloisInc/llvm-pretty-bc-parser/issues/248). Unfortunately, this only covers the case where a known metadata type is parsed, but with a different number of records than expected. If you encounter an entirely new metadata record code, on the other hand, then `llvm-pretty-bc-parser` will still treat that as a hard error.

As an example, consider what happens if you parse a bitcode file produced from the `disasm-test/tests/difixedpointtype.ll` test case added in #354, but using an `llvm-pretty-bc-parser` commit that predates #354 (e.g., 6f49d2c295a9f5676e798a4f189da0a274691134). This commit lacks support for parsing metadata record code 49 (`METADATA_FIXED_POINT_TYPE`), and this causes `llvm-pretty-bc-parser` to fail with the following error:

```
λ> :m + Text.LLVM.PP Data.LLVM.BitCode
λ> parseBitCodeFromFileWithWarnings "test.bc" >>= either (putStrLn . formatError) (print . ppLLVM llvmVlatest ppModule . fst)
unknown record code: 49
from:
	METADATA_BLOCK
	METADATA_BLOCK_ID
	value symbol table
	MODULE_BLOCK
	Bitstream
```

Like with #248, we should downgrade this from an error to a warning.

**Comment by RyanGlScott** [comment]
Unfortunately, this is not as straightforward as I was expecting. I tried something like this:

<details>

```diff
diff --git a/src/Data/LLVM/BitCode/IR/Metadata.hs b/src/Data/LLVM/BitCode/IR/Metadata.hs
index 6d8f82d..0e4660e 100644
--- a/src/Data/LLVM/BitCode/IR/Metadata.hs
+++ b/src/Data/LLVM/BitCode/IR/Metadata.hs
@@ -455,6 +455,11 @@ parseMetadataEntry vt mt pm (fromEntry -> Just r) =
             (Just . ValMdValue . Typed (PrimType (Integer 64)) . ValInteger)
             (parseField r n numeric)

+      mdNotImplemented :: Parse PartialMetadata
+      mdNotImplemented = do
+        addParseWarning $ UnsupportedMetadataRecordCode (recordCode r)
+        pure pm
+
   -- Note: the parsing cases below use a Monadic coding style, as opposed to an
   -- Applicative style (as was originally used) for performance reasons:
   -- Applicative record construction has quadratic size and corresponding
@@ -1194,16 +1199,16 @@ parseMetadataEntry vt mt pm (fromEntry -> Just r) =
         (addDebugInfo isDistinct (DebugInfoLabel dil)) pm

     41 -> label "METADATA_STRING_TYPE" $ do
-      notImplemented
+      mdNotImplemented

     -- Codes 42 and 43 are reserved for Fortran array–specific debug info, see
     -- https://github.com/llvm/llvm-project/blob/4681f6111e655057f5015564a9bf3705f87495bf/llvm/include/llvm/Bitcode/LLVMBitCodes.h#L348-L349

     44 -> label "METADATA_COMMON_BLOCK" $ do
-      notImplemented
+      mdNotImplemented

     45 -> label "METADATA_GENERIC_SUBRANGE" $ do
-      notImplemented
+      mdNotImplemented

     46 -> label "METADATA_ARG_LIST" $ do
       cxt <- getContext
@@ -1243,7 +1248,9 @@ parseMetadataEntry vt mt pm (fromEntry -> Just r) =
       return $! updateMetadataTable
         (addDebugInfo isDistinct (DebugInfoSubrangeType disrt)) pm

-    code -> fail ("unknown record code: " ++ show code)
+    code -> do
+      addParseWarning $ UnknownMetadataRecordCode code
+      pure pm

 parseMetadataEntry _ _ pm (abbrevDef -> Just _) =
   return pm
diff --git a/src/Data/LLVM/BitCode/Parse.hs b/src/Data/LLVM/BitCode/Parse.hs
index 732b384..b0e40c1 100644
--- a/src/Data/LLVM/BitCode/Parse.hs
+++ b/src/Data/LLVM/BitCode/Parse.hs
@@ -806,6 +806,12 @@ data ParseWarning
     -- the @[String]@ is the stack trace at the point where the warning was
     -- emitted.
     InvalidMetadataRecordSize !Int !MetadataRecordSizeRange ![String]
+  | -- | The parser encountered a metadata record code that it is aware of, but
+    -- does not know how to parse, where the 'Int' is the record code value.
+    UnsupportedMetadataRecordCode !Int
+  | -- | The parser encountered a metadata record code that it is not aware of,
+    -- where the 'Int' is the record code value.
+    UnknownMetadataRecordCode !Int
   deriving Show

 -- | The expected size of a metadata record.
@@ -838,6 +844,10 @@ ppParseWarning pw =
         [ "Invalid record size:" PP.<+> PP.pPrint len
         , expectedSizeMsg range
         ] ++ map PP.text (formatContext cxt)
+    UnsupportedMetadataRecordCode code ->
+      "Unsupported record code:" PP.<+> PP.pPrint code
+    UnknownMetadataRecordCode code ->
+      "Unknown record code:" PP.<+> PP.pPrint code
   where
     expectedSizeMsg :: MetadataRecordSizeRange -> PP.Doc
     expectedSizeMsg range =
@@ -863,12 +873,14 @@ ppParseWarnings warnings
         (\warning ->
           PP.nest 4 $ PP.vcat [ppParseWarning warning, ""])
         (F.toList warnings) ++
-      [supportMsg | any isInvalidMetadataRecordSize warnings]
+      [supportMsg | any isInvalidMetadata warnings]
   where
-    isInvalidMetadataRecordSize :: ParseWarning -> Bool
-    isInvalidMetadataRecordSize pw =
+    isInvalidMetadata :: ParseWarning -> Bool
+    isInvalidMetadata pw =
       case pw of
         InvalidMetadataRecordSize {} -> True
+        UnsupportedMetadataRecordCode {} -> True
+        UnknownMetadataRecordCode {} -> True

     supportMsg :: PP.Doc
     supportMsg =
```
</details>

After applying these changes, however, if I try to parse a bitcode file produced from `disasm-test/tests/difixedpointtype.ll`, I get a different error:

```
λ> :m + Text.LLVM.PP Data.LLVM.BitCode
λ> parseBitCodeFromFileWithWarnings "test.bc" >>= either (putStrLn . formatError) (print . ppLLVM llvmVlatest ppModule . fst)
bad forward reference to value: 21
additional details:
Bad reference into a value table
with call stack:
CallStack (from HasCallStack):
  forwardRef, called at src/Data/LLVM/BitCode/IR/Metadata.hs:163:28 in llvm-pretty-bc-parser-0.6.0.0.99-inplace:Data.LLVM.BitCode.IR.Metadata

from:
	METADATA_COMPILE_UNIT
	METADATA_BLOCK
	METADATA_BLOCK_ID
	value symbol table
	MODULE_BLOCK
	Bitstream
```

I suspect that skipping over parsing the `METADATA_FIXED_POINT_TYPE` metadata caused `llvm-pretty-bc-parser`'s internal bookkeeping to become out of sync somehow, although I have not looked deeply into why this happens.

**Comment by kquick** [comment]
Metadata nodes can reference each other.  I suspect we will need to insert some sort of "null" entry for the MD nodes we cannot parse to act as a placeholder.
